File: sdks/php/src/Schema.php
<?php

namespace Autonoma\Sdk;

use Autonoma\Sdk\Types\FieldInfo;
use Autonoma\Sdk\Types\ModelInfo;
use Autonoma\Sdk\Types\SchemaInfo;

/**
 * Build the SDK's wire-shape schema from registered factories.
 *
 * The dashboard's discover response carries a schema block that lists every
 * model the host can create along with each model's fields. With factories,
 * this comes from each factory's inputFields array.
 */
class Schema
{
    /**
     * Valid SDK type strings for input fields.
     */
    private const VALID_TYPES = [
        'string', 'integer', 'number', 'boolean',
        'timestamp', 'date', 'uuid', 'json',
    ];

    /**
     * Convert OrgMember to org_member for cosmetic tableName.
     */
    private static function camelToSnake(string $name): string
    {
        $out = '';
        for ($i = 0; $i < strlen($name); $i++) {
            $ch = $name[$i];
            if (ctype_upper($ch) && $i > 0 && !ctype_upper($name[$i - 1])) {
                $out .= '_';
            }
            $out .= strtolower($ch);
        }
        return $out;
    }

    /**
     * Build FieldInfo list from a factory's inputFields, prepending a synthetic id field.
     *
     * @param FieldInfo[] $inputFields
     * @return FieldInfo[]
     */
    private static function buildModelFields(array $inputFields): array
    {
        $fields = [
            new FieldInfo(
                name: 'id',
                type: 'string',
                isRequired: false,
                isId: true,
                hasDefault: true,
            ),
        ];

        foreach ($inputFields as $field) {
            $fields[] = $field;
        }

        return $fields;
    }

    /**
     * Build the SDK's discover-time schema from registered factories.
     *
     * @param array<string, \Autonoma\Sdk\Types\FactoryDefinition> $factories
     */
    public static function buildSchemaFromFactories(array $factories, string $scopeField): SchemaInfo
    {
        $models = [];

        foreach ($factories as $entity => $factory) {
            if (empty($factory->inputFields)) {
                throw new \RuntimeException(
                    "Factory \"{$entity}\" has no inputFields. " .
                    "Every factory must declare inputFields in defineFactory(...)."
                );
            }

            $models[] = new ModelInfo(
                name: $entity,
                tableName: self::camelToSnake($entity),
                fields: self::buildModelFields($factory->inputFields),
            );
        }

        return new SchemaInfo(
            models: $models,
            edges: [],
            relations: [],
            scopeField: $scopeField,
        );
    }

    /**
     * Serialise a SchemaInfo to the JSON shape the dashboard expects.
     *
     * @return array<string, mixed>
     */
    public static function schemaToWire(SchemaInfo $schema): array
    {
        return [
            'models' => array_map(fn(ModelInfo $m) => [
                'name' => $m->name,
                'tableName' => $m->tableName,
                'fields' => array_map(fn(FieldInfo $f) => [
                    'name' => $f->name,
                    'type' => $f->type,
                    'isRequired' => $f->isRequired,
                    'isId' => $f->isId,
                    'hasDefault' => $f->hasDefault,
                ], $m->fields),
            ], $schema->models),
            'edges' => [],
            'relations' => [],
            'scopeField' => $schema->scopeField,
        ];
    }
}

File: sdks/rust/src/schema.rs
//! Build the SDK's wire-shape schema from registered factories.
//!
//! The dashboard's `discover` response carries a `schema` block that lists
//! every model the host can create, along with each model's fields. With
//! this design it comes from each factory's `input_fields` (a `Vec<FieldDef>`).

use serde_json::{json, Value};

use crate::types::{FactoryRegistry, FieldInfo, ModelInfo, SchemaInfo};

/// Convert `OrgMember` to `org_member` for cosmetic `tableName`.
fn camel_to_snake(name: &str) -> String {
    let mut out = String::new();
    for (i, ch) in name.chars().enumerate() {
        if ch.is_uppercase() && i > 0 {
            let prev = name.as_bytes()[i - 1];
            if !prev.is_ascii_uppercase() {
                out.push('_');
            }
        }
        out.push(ch.to_ascii_lowercase());
    }
    out
}

/// Build the SDK's discover-time schema from registered factories.
///
/// `edges` and `relations` are emitted as empty lists. They were populated
/// from FK introspection in the old design; here the create payload's
/// `_alias`/`_ref` graph carries equivalent information at request time.
pub fn build_schema_from_factories(
    factories: &FactoryRegistry,
    scope_field: &str,
) -> SchemaInfo {
    let mut models: Vec<ModelInfo> = Vec::new();

    for (entity, factory) in factories {
        // Every model gets a synthetic `id` field at the head.
        let mut fields: Vec<FieldInfo> = vec![FieldInfo {
            name: "id".to_string(),
            field_type: "string".to_string(),
            is_required: false,
            is_id: true,
            has_default: true,
        }];

        for fd in &factory.input_fields {
            fields.push(FieldInfo {
                name: fd.name.clone(),
                field_type: fd.field_type.clone(),
                is_required: fd.required,
                is_id: false,
                has_default: !fd.required,
            });
        }

        models.push(ModelInfo {
            name: entity.clone(),
            table_name: camel_to_snake(entity),
            fields,
        });
    }

    SchemaInfo {
        models,
        edges: vec![],
        relations: vec![],
        scope_field: scope_field.to_string(),
    }
}

/// Serialise a `SchemaInfo` to the JSON shape the dashboard expects.
pub fn schema_to_wire(schema: &SchemaInfo) -> Value {
    json!({
        "models": schema.models.iter().map(|m| {
            json!({
                "name": m.name,
                "tableName": m.table_name,
                "fields": m.fields.iter().map(|f| {
                    json!({
                        "name": f.name,
                        "type": f.field_type,
                        "isRequired": f.is_required,
                        "isId": f.is_id,
                        "hasDefault": f.has_default,
                    })
                }).collect::<Vec<_>>(),
            })
        }).collect::<Vec<_>>(),
        "edges": schema.edges.iter().map(|e| {
            json!({
                "from": e.from_model,
                "to": e.to_model,
                "localField": e.local_field,
                "foreignField": e.foreign_field,
                "nullable": e.nullable,
            })
        }).collect::<Vec<_>>(),
        "relations": schema.relations.iter().map(|r| {
            json!({
                "parentModel": r.parent_model,
                "childModel": r.child_model,
                "parentField": r.parent_field,
                "childField": r.child_field,
            })
        }).collect::<Vec<_>>(),
        "scopeField": schema.scope_field,
    })
}
